akka-remote/src/main/scala/akka/remote/artery/Codecs.scala

```diff
@@ -129,15 +144,46 @@ private[remote] class Encoder(
         } finally {
           outboundEnvelope match {
             case r: ReusableOutboundEnvelope ⇒ outboundEnvelopePool.release(r)
-            case _                           ⇒
+            case _                           ⇒ // no need to release it
           }
         }
-
       }
 
       override def onPull(): Unit = pull(in)
 
       /**
+       * Renders metadata into `headerBuilder`.
+       *
+       * Replace all AnyRef's that were passed along with the [[OutboundEnvelope]] into their [[ByteString]] representations,
+       * by calling `remoteMessageSent` of each enabled instrumentation. If `context` was attached in the envelope it is passed
+       * into the instrument, otherwise it receives an OptionVal.None as context, and may still decide to attach rendered
+       * metadata by returning it.
+       */
+      private def applyAndRenderRemoteMessageSentMetadata(instruments: Vector[RemoteInstrument], envelope: OutboundEnvelope, headerBuilder: HeaderBuilder) = {
+        // apply all outgoing instruments, 0 and 1 instrumentation has fast-path
+        if (instruments.nonEmpty) {
+          val n = instruments.length
+          serializedMetadatas.clear()
+
+          var i = 0
+          while (i < n) {
+            val instrument = instruments(i)
+            val instrumentId = instrument.identifier
+
+            val context: OptionVal[AnyRef] = envelope.metadata(instrumentId)
+            instrument.remoteMessageSent(envelope.recipient, envelope.message, envelope.sender, context) match {
+              case OptionVal.Some(renderedMetadata) ⇒ serializedMetadatas.set(instrumentId, renderedMetadata)
+              case _                                ⇒ // no metadata to attach
+            }
+
+            i += 1
+          }
```

Comment:
I'm not sure... you mean giving a ByteBuffer to the SPI etc right?
Technically most efficient but people can screw up stuff a lot then hm.

I guess it's a crappy instrument then.

Any opinion here @bantonsson ?